chip8.cpp

```diff
@@ -97,39 +102,239 @@ void Chip8::decodeAndExecute(uint16_t opcode) {
+                } else {
+                    registers[0xF] = 0;
+                }
+                registers[registerVX] -= registers[registerVY];
+                programCounter += 2;
+            }
+            if(N == 0x0006) {
+                registers[registerVX] = registers[registerVY];
+                bool bit = registers[registerVX] & 1;
+                registers[registerVX] >>= 1;
+
+                if(bit) {
+                    registers[0xF] = 1;
+                } else {
+                    registers[0xF] = 0;
+                }
+                programCounter += 2;
+            }
+            if(N == 0x0007) {
+                if(registers[registerVX] < registers[registerVY]) {
+                    registers[0xF] = 1;
+                } else {
+                    registers[0xF] = 0;
+                }
+                registers[registerVX] = registers[registerVY] - registers[registerVX];
+                programCounter += 2;
+            }
+            if(N == 0x000E) {
+                registers[0xF] = registers[registerVX] >> 7;
+                registers[registerVX] <<= 1;
+                programCounter += 2;
+            }
             break;
         case 0x9:
+            if(registers[registerVX] != registers[registerVY]) {
+                programCounter += 2;
+            }
+            programCounter += 2;
             break;
         case 0xA:
+            index = NNN;
+            programCounter += 2;
             break;
         case 0xB:
+            programCounter = NNN + registers[0];
             break;
         case 0xC:
+            registers[registerVX] = randByte(randGen) & NN;
+            programCounter += 2;
             break;
         case 0xD:
+            unsigned short pixel;
+            x = registers[registerVX] % 64;
+            y = registers[registerVY] % 32;
+            height = N;
+
+            registers[0xF] = 0;
+            for(int yAxis = 0; yAxis < height; yAxis++){
+                pixel = memory[index + yAxis];
+                for (int xAxis = 0; xAxis < 8; xAxis++) {
+                    if((pixel & (0x80 >> xAxis)) != 0) {
+                        if(video[(x + xAxis + ((y + yAxis) *64))] == 1) {
+                            video[0xF] = 1;
+                        }
+                        video[x+xAxis + ((y + yAxis) * 64)] ^= 1;
+                    }
+                }
+            }
+            programCounter += 2;
+            BOOST_LOG_TRIVIAL(info) << "Coordinates Drawn";
             break;
         case 0xE:
+            if(NN == 0x009E) {
+                if (keypad[registers[registerVX]] != 0) {
+                    programCounter += 2;
+                } 
+                programCounter += 2;
+            }
+            if(NN == 0x00A1) {
+                if (keypad[registers[registerVX]] == 0) {
+                    programCounter += 2;
+                }
+                programCounter += 2;
+            }
             break;
         case 0xF:
```

Comment:
There is a bug that is caused in this instruction and fails the test_opcode.ch8 rom

Reply:
[8a2464a](https://github.com/ahchaprazian/Chip8-Emulator/pull/2/commits/8a2464acbc68f1986b8bd7d2bdbe357955bf7e30) Failing instructions were 0xF033 and 0xF055 because 0xF055 needed a <= in the for loop instead of <. The failure was because one case in the registers was not written into the memory.